### PR TITLE
Add `normalize` method to blocks (revisited)

### LINKED ISSUE #9738: "normalize" mechanism for StreamField blocks
### Is your proposal related to a problem?

When writing to a StreamField block, e.g. `page.body = [('heading', "Hello world!")]`, or setting a `default` value, we expect the block value to be passed in the block's native data type. This data type is often undocumented and non-obvious:

* for RichTextBlock, it's a `wagtail.rich_text.RichText` object, not a string
* for EmbedBlock, it's an EmbedValue object
* for StreamBlock, it's a `StreamValue` instance. (The list-of-tuples representation is only recognised by `StreamField.to_python`, not StreamBlock, which means it's only valid for the top level of the stream, not nested StreamBlocks.)
* For ListBlock, it used to be a Python list, but is now a `ListValue` instance as of Wagtail 4.1.

StreamValue (and possibly ListValue too, now) is especially troublesome, as the constructor needs to be passed a reference to the StreamBlock, which is hard to get hold of, and impossible in the case of setting a `default` (because it would have to be a reference to the StreamBlock that you're still in the middle of defining).

In these situations, we should follow Postel's Law and accept "close enough" data types whenever a block value is being supplied from user code: a string for RichTextBlock, a list-of-tuples for StreamBlock and so on.

### Describe the solution you'd like

The `Block` class should implement a `normalize(value)` method that accepts an input value in a range of allowable types, and returns that value in the block's native type. This differs from existing methods such as `to_python` in that it's not aiming to convert data between representations - just standardising data that "should" be in the native type. We would then run any user-supplied data through `normalize` before using it in the block's internal logic.

The default implementation in `Block` would simply return `value` unchanged, while blocks that act as containers for child blocks will recursively call `normalize` on the child values along with doing whatever conversion they require for themselves.

### Additional context

If we were to move the "sniffing various types to decide what to do" logic from StreamField.to_python to StreamBlock.normalize, I believe this would remove the remaining blocker to supporting other kinds of block besides StreamBlock as the top-level block of a StreamField, providing a solution to #2048.

### PR DESCRIPTION
**Pull request by gasman**
Rework of #9746 with StreamBlock's `normalize` and `to_python` methods unified into one method that handles all incoming formats, plus some extra unit tests. Fixes #9738 

Not sure if we need additional documentation - the improvements here only really show up in more complex nested structures than the examples in the documentation, and those are things that logically _should_ have worked by following and extending the code examples, but didn't.

### CODE REVIEW

**Comment by gasman** [comment]
Requesting review from @laymonage - I'm happy with @jams2's commits here, but would welcome another pair of eyes on my additions :-)

**Review by laymonage — APPROVED** [review]
I'm not entirely familiar with StreamField's internals, but looking at the tests, this makes sense to me.

The back-and-forth between `normalize` vs `to_python` threw me off for a moment. The discussion in the previous PR cleared it up, though.

The final diff is much easier to follow, since we're basically moving `StreamField.to_python`'s logic to `StreamBlock.to_python` and using the newly-added `normalize()` to tidy up the process. Thanks! 👍

File: wagtail/blocks/static_block.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def get_admin_text(self):
     def value_from_datadict(self, data, files, prefix):
         return None
 
-    def normalize(self, _):
+    def normalize(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/blocks/static_block.py]: Good question! Reverted in 1cbe206362604e1d0c03c001b6a7a51d9e3ec012.

**Comment by gasman** [comment]
Merged in abd018b9262a5d915986a287331e5e4d81b1659e

**Pull request closed without merging**